src/server.c

```diff
@@ -944,32 +965,304 @@ static int on_client_pointer_event(struct nvnc_client* client)
+	if (client->server->extcb_provide_msg.buffer)
+		flags |= RFB_EXTCB_FORMAT_TEXT;
+	msg.flags = htonl(flags);
+
+	stream_write(client->net_stream, &msg, sizeof(msg), NULL, NULL);
+}
+
+static void send_extcb_provide(struct nvnc_client* client)
+{
+	assert(client->server->extcb_provide_msg.buffer);
+
+	struct rfb_extcb_msg msg;
+
+	msg.type = RFB_SERVER_TO_CLIENT_SERVER_CUT_TEXT;
+	msg.length = htonl(-(4 + client->server->extcb_provide_msg.length));
+	msg.flags = htonl(
+			RFB_EXTCB_ACTION_PROVIDE |
+			RFB_EXTCB_FORMAT_TEXT);
+
+	stream_write(client->net_stream, &msg, sizeof(msg), NULL, NULL);
+	stream_write(client->net_stream,
+			client->server->extcb_provide_msg.buffer,
+			client->server->extcb_provide_msg.length,
+			NULL, NULL);
+}
+
+static char* crlf_to_lf(const char* src, size_t len)
+{
+	char* lf_buf = NULL;
+	size_t lf_size;
+
+	/* caller will read as a null-terminated string, so make the last byte 0 */
+	lf_size = 1;
+
+	const char* in;
+	size_t in_len;
+
+	/* determine lf_size */
+	in = src;
+	in_len = len;
+	while (in_len > 0) {
+		if (*in != '\r') {
+			lf_size++;
+			in++;
+			in_len--;
+			continue;
+		}
+
+		if ((in_len == 0) || (*(in + 1) != '\n'))
+			lf_size++;
+
+		in++;
+		in_len--;
 	}
+
+	lf_buf = calloc(1, lf_size);
+	if (!lf_buf)
+		return NULL;
+
+	char* out = lf_buf;
+	in = src;
+	in_len = len;
+	while (in_len > 0) {
+		if (*in != '\r') {
+			*out++ = *in++;
+			in_len--;
+			continue;
+		}
+
+		if ((in_len == 0) || (*(in + 1) != '\n'))
+			*out++ = '\n';
+
+		in++;
+		in_len--;
+	}
+
+	return lf_buf;
 }
 
-static int on_client_cut_text(struct nvnc_client* client)
+static void process_client_extcb_provide(struct nvnc_client* client,
+		unsigned char* zlib_data, size_t zlib_len)
 {
-	struct rfb_cut_text_msg* msg =
-	        (struct rfb_cut_text_msg*)(client->msg_buffer +
-	                                   client->buffer_index);
+	int rc;
+
+	z_stream zs;
+	zs.zalloc = NULL;
+	zs.zfree = NULL;
+	zs.opaque = NULL;
+	zs.avail_in = 0;
+	zs.next_in = NULL;
+	if (inflateInit(&zs) != Z_OK)
+		return;
+
+	uint32_t inflate_len;
+
+	zs.avail_in = zlib_len;
+	zs.next_in = zlib_data;
+	zs.avail_out = 4;
+	zs.next_out = (unsigned char*)&inflate_len;
+	rc = inflate(&zs, Z_SYNC_FLUSH);
+	if (rc != Z_OK) {
+		nvnc_log(NVNC_LOG_WARNING, "Failed to inflate client's clipboard text: %p (ref %d)",
+				client, client->ref);
+		inflateEnd(&zs);
+		return;
+	}
+	inflate_len = ntohl(inflate_len);
+
+	/* null terminator is not wanted */
+	if (inflate_len <= 1) {
+		nvnc_log(NVNC_LOG_DEBUG, "Client sent empty clipboard update: %p (ref %d)",
+				client, client->ref);
+		inflateEnd(&zs);
+		return;
+	}
+	inflate_len--;
+
+	unsigned char* inflate_buf = malloc(inflate_len);
+	if (!inflate_buf) {
+		nvnc_log(NVNC_LOG_ERROR, "OOM: %m");
+		inflateEnd(&zs);
+		return;
+	}
+
+	zs.avail_out = inflate_len;
+	zs.next_out = inflate_buf;
+	rc = inflate(&zs, Z_SYNC_FLUSH);
+	/* it won't return Z_STREAM_END because we only allow it to read up
```

Comment:
`inflateEnd` can be hoisted up here.

I've a feeling that not getting `Z_STREAM_END` here might be a client bug. As far as I understand the spec, the null byte should be included in the message and `size` in the "provide" header should be the full size of the message.

This needs further investigation.

Reply:
It is there, but I chose not to inflate the last byte where the null terminator is (`inflate_len--;` before copying to `avail_out`), so I never finish the inflate.

Reply:
Ah, I see. But why skip it? It's a good sanity check to make sure that the stream ends where its supposed to.